core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java

```diff
@@ -2911,13 +2917,239 @@ else if (isStringType(type)) {
+                // - all other values are cast to VARCHAR
+
+                // if the value expression is a JSON-returning function, there should be an explicit or implicit input format (spec p.817)
+                // JSON-returning functions are: JSON_OBJECT, JSON_OBJECTAGG, JSON_ARRAY, JSON_ARRAYAGG and JSON_QUERY
+                if ((value instanceof JsonQuery ||
+                        value instanceof JsonObject ||
+                        value instanceof JsonArray) && // TODO add JSON_OBJECTAGG, JSON_ARRAYAGG when supported
+                        format.isEmpty()) {
+                    format = Optional.of(JsonFormat.JSON);
+                }
+
+                Type valueType = process(value, context);
+
+                if (format.isPresent()) {
+                    withoutUniqueKeysEnforced = true;
+                    // resolve function to read the value as JSON
+                    ResolvedFunction inputFunction = getInputFunction(valueType, format.get(), value);
+                    Type expectedValueType = inputFunction.getSignature().getArgumentType(0);
+                    coerceType(value, valueType, expectedValueType, "value passed to JSON_OBJECT function");
+                    jsonInputFunctions.put(NodeRef.of(value), inputFunction);
+                    valueType = JSON_2016;
+                }
+                else {
+                    if (isStringType(valueType)) {
+                        if (!isCharacterStringType(valueType)) {
+                            throw semanticException(NOT_SUPPORTED, value, "Unsupported type of value passed to JSON_OBJECT function: %s", valueType.getDisplayName());
+                        }
+                    }
+
+                    if (!isStringType(valueType) && !isNumericType(valueType) && !valueType.equals(BOOLEAN)) {
+                        if (!typeCoercion.canCoerce(valueType, VARCHAR)) {
+                            throw semanticException(INVALID_FUNCTION_ARGUMENT, value, "Invalid type of JSON object value: %s", valueType.getDisplayName());
+                        }
+                        coerceType(value, valueType, VARCHAR, "value passed to JSON_OBJECT function");
+                        valueType = VARCHAR;
+                    }
+                }
+
+                valueFields.add(new RowType.Field(Optional.empty(), valueType));
+            }
+
+            RowType keysRowType = JSON_NO_PARAMETERS_ROW_TYPE;
+            RowType valuesRowType = JSON_NO_PARAMETERS_ROW_TYPE;
+            if (!node.getMembers().isEmpty()) {
+                keysRowType = RowType.from(keyFields.build());
+                valuesRowType = RowType.from(valueFields.build());
+            }
+
+            if (node.isUniqueKeys()) {
+                if (withoutUniqueKeysEnforced) {
+                    throw semanticException(NOT_SUPPORTED, node, "WITH UNIQUE KEYS behavior is not supported for JSON_OBJECT function when input expression has FORMAT");
+                }
+            }
```

Comment:
would it make sense to check and maybe throw already above, in place where you are setting `withoutUniqueKeysEnforced = true;`